crates/bifrost/src/log_chain_extender.rs

```diff
@@ -0,0 +1,180 @@
+pub enum ExtendLogChainError {
+    #[error("unknown log '{0}'")]
+    UnknownLogId(LogId),
+    #[error("segment index found in metadata does not match expected {expected}!={found}")]
+    SegmentMismatch {
+        expected: SegmentIndex,
+        found: SegmentIndex,
+    },
+    #[error(transparent)]
+    LogsBuilder(Arc<BuilderError>),
+    #[error(transparent)]
+    Shutdown(#[from] ShutdownError),
+    #[error("failed writing logs to metadata store: {0}")]
+    MetadataStore(Arc<ReadWriteError>),
+    #[error("Logs have not been initialized yet.")]
+    EmptyLogs,
+}
+
+impl From<ReadWriteError> for ExtendLogChainError {
+    fn from(value: ReadWriteError) -> Self {
+        ExtendLogChainError::MetadataStore(Arc::new(value))
+    }
+}
+
+pub(super) struct ExtendLogChain {
+    pub log_id: LogId,
+    pub last_segment_index: SegmentIndex,
+    pub base_lsn: Lsn,
+    pub provider: ProviderKind,
+    pub params: LogletParams,
+    pub response_tx: Option<oneshot::Sender<Result<(), ExtendLogChainError>>>,
+}
+
+impl ExtendLogChain {
+    fn fail(&mut self, err: ExtendLogChainError) {
+        if let Some(response_tx) = self.response_tx.take() {
+            // ignore if the receiver disappeared
+            let _ = response_tx.send(Err(err));
+        }
+    }
+
+    fn complete(&mut self) {
+        if let Some(response_tx) = self.response_tx.take() {
+            // ignore if the receiver disappeared
+            let _ = response_tx.send(Ok(()));
+        }
+    }
+}
+
+/// Component which coalesces multiple log-chain updates into a single [`Logs`] update. It works
+/// by draining all available [`ExtendLogChain`] commands and applying them to the current logs
+/// configuration using a read-modify-write metadata operation. A log chain can only be extended if
+/// the last segment index equals the value specified by the [`ExtendLogChain`] command.
+pub struct LogChainExtender {
+    inner: Arc<BifrostInner>,
+    extend_log_chain_rx: ExtendLogChainReceiver,
+}
+
+impl LogChainExtender {
+    pub fn new(inner: Arc<BifrostInner>, extend_log_chain_rx: ExtendLogChainReceiver) -> Self {
+        Self {
+            inner,
+            extend_log_chain_rx,
+        }
+    }
+
+    pub async fn run(self) -> anyhow::Result<()> {
+        trace!("Bifrost log chain extender started");
+
+        cancellation_token()
+            .run_until_cancelled(self.run_inner())
+            .await
+            .ok_or(ShutdownError)?;
+
+        Ok(())
+    }
+
+    pub async fn run_inner(mut self) {
+        let mut buffer = Vec::new();
+
+        // await the first extend log chain command
+        while let Some(extend_log_chain) = self.extend_log_chain_rx.recv().await {
+            buffer.push(extend_log_chain);
+
+            // drain all available extend log chain commands
+            while let Ok(extend_log_chain) = self.extend_log_chain_rx.try_recv() {
+                buffer.push(extend_log_chain);
+            }
```

Comment:
would `recv_many(&mut buffer, 128)` be a good candidate here? it would also put an upper bound to the number of operations we want to buffer.